File: Algorithm_problem_solving/Baek-joon/2206/2206.py
# https://www.acmicpc.net/problem/2206
import sys
from collections import deque
input = sys.stdin.readline

way = (-1, 0), (1, 0), (0, -1), (0, 1)

def BFS():
    # initial setting
    que = deque()
    que.append([1, 1, 1, False])
    visited = {}

    # Repeat until que is empty
    while que:
        # 벽은 부순적이 있느ㄴ지 없는지 저장하는 변수
        y, x, weight, broken = que.popleft()
        if visited.get((y, x, broken)):
            continue
        # 벽을 만났을때, 지금까지 벽을 부순 적이 없다면 벽을 부순다.
        if board[y][x] == 1:
            if not broken:
                broken=True
            else:
                continue
        # 목적지에 도달하였다면 시간을 출력하고 return
        if x == M and y == N:
            print(weight)
            return
        # 벽을 부수고 방문한 경우와 암 부수고 그 장소에 방문한 경우는 다르므로 broken변수도 같이 저장
        visited[(y, x, broken)] = True
        for next in way:
            next_x = x + next[0]
            next_y = y + next[1]
            if next_x <= 0 or next_x > M or next_y <= 0 or next_y > N:
                continue
            que.append([next_y, next_x, weight+1, broken])
    # 모든 반복문이 끝났는데 목적지에 도달하지 못했다면 -1 출력
    print(-1)

if __name__ == "__main__":
    N, M = map(int, input().strip().split())
    board = [[0]*(M+1) for i in range(N+1)]
    for i in range(N):
        temp = input().strip()
        for j in range(M):
            board[i+1][j+1] = int(temp[j])
    BFS()

File: Algorithm_problem_solving/Baek-joon/2206/2206.cpp
#include <iostream>
#include <string>
#include <queue>

using namespace std;

int board[1000][1000];
bool visited[1000][1000][2];

struct Player
{
    int y;
    int x;
    int moveCnt;
    bool broken;
};

class Solution
{
private:
    int way[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    int n, m;
    // int (*board)[1000];

public:
    Solution(int n, int m /*, int board[][1000]*/)
    {
        this->n = n;
        this->m = m;
        // this->board = board;
    }

    int getMinDistance()
    {
        queue<Player> que;
        Player curr{0, 0, 1, false};

        que.push(curr);
        visited[curr.y][curr.x][0] = true;

        int nextY, nextX;
        while (!que.empty())
        {
            curr = que.front();
            que.pop();
            if (curr.y == n - 1 && curr.x == m - 1)
            {
                return curr.moveCnt;
            }
            for (int i = 0; i < 4; i++)
            {
                nextY = curr.y + way[i][0];
                nextX = curr.x + way[i][1];
                if (nextY < 0 || nextX < 0 || nextY >= n || nextX >= m)
                    continue;
                else if (visited[nextY][nextX][curr.broken ? 1 : 0])
                    continue;
                else if (board[nextY][nextX] == 1)
                {
                    if (curr.broken)
                        continue;
                    else
                    {
                        que.push(Player{nextY, nextX, curr.moveCnt + 1, true});
                        visited[nextY][nextX][1] = true;
                    }
                }
                else
                {
                    que.push(Player{nextY, nextX, curr.moveCnt + 1, curr.broken});
                    visited[nextY][nextX][curr.broken ? 1 : 0] = true;
                }
            }
        }
        return -1;
    }
};

int main(void)
{
    int n, m;

    scanf("%d %d", &n, &m);

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            scanf("%1d", &board[i][j]);
        }
    }

    Solution *solution = new Solution(n, m);
    cout << solution->getMinDistance() << endl;
    return 0;
}